File: SoftwareDesign/Week3/Example2/Sample.java
abstract class Human {
   abstract public void help();
}

interface Helper {
    public void help();
}

class Man implements Helper {
    public void help() {
        System.out.println("man helping");
    }
}
class Woman implements Helper {
    public void help() {
        System.out.println("woman helping");
    }
}
class Elephant implements Helper {
    public void help() {
        System.out.println("elephant can help quite a bit");
    }
}


public class Sample {
    public static void seekHelpFrom(Human helper) {
        helper.help();
    }
    public static void main(String[] args) {
        Man man = new Man();
        Woman woman = new Woman();
        seekHelpFrom(man);
        seekHelpFrom(woman);

        Elephant hathi = new Elephant();
        seekHelpFrom(hathi);
    }
}

File: SoftwareDesign/Week3/Example3/sample.py
class Man:
    def __init__(self):
        pass

    def help(self):
        print("man helping...")

class Woman:
    def __init__(self):
        pass

    def help(self):
        print("woman helping...")

class Elephant:
    def __init(self):
        pass

    def help(self):
        print("elephant helping...")

class Fox:
    def __init__(self):
        pass


def seekHelp(helper):
    # python won't check if it's a method or field ):
    if ("help" in dir(helper)):
        helper.help()
    else:
        print("Are you kidding...")

def main():
    bob = Man()
    sarah = Woman()
    hathi = Elephant()
    fox = Fox()
    seekHelp(bob)
    seekHelp(sarah)
    seekHelp(hathi)
    seekHelp(fox)


main()